### ISSUE TITLE
sketch on face extrude semantics

### ISSUE THREAD

**Issue by jessfraz** [KittyCAD/modeling-app#2395]
from user of app

some of these might be behaving as intended until we have subtract

```
fn rect = (sk, pos, l, w) => {
  const rr = sk
    |> startProfileAt([pos[0] - (w / 2), pos[1] - (l / 2)], %)
    |> lineTo([pos[0] + w / 2, pos[1] - (l / 2)], %, 'edge1')
    |> lineTo([pos[0] + w / 2, pos[1] + l / 2], %, 'edge2')
    |> lineTo([pos[0] - (w / 2), pos[1] + l / 2], %, 'edge3')
    |> close(%, "edge4")
  return rr
}

// ✅
const p1 = rect(startSketchOn('XY'), [0, 0], 50, 50)
  |> extrude(10, %)
const p1_1 = rect(startSketchOn(p1, "end"), [0, 0], 20, 20)
  |> extrude(10, %)

// ✅
const p2 = rect(startSketchOn('XY'), [60, 0], 50, 50)
  |> extrude(10, %)
const p2_1 = rect(startSketchOn(p2, "end"), [60, 0], 20, 20)
  |> extrude(-10, %)

// ❌ negative extrude through more than width of extrusion
const p5 = rect(startSketchOn('XY'), [120, 0], 50, 50)
  |> extrude(10, %)
const p5_1 = rect(startSketchOn(p5, "end"), [120, 0], 20, 20)
  |> extrude(-20, %)

// ❌ extrude with edge coincident boundary this shows red lines
const p3 = rect(startSketchOn('XY'), [0, 60], 50, 50)
  |> extrude(10, %)
const p3_1 = rect(startSketchOn(p3, "end"), [0, 45], 20, 20)
  |> extrude(10, %)

// ❌ negative extrude with edge coincident boundary (this doesn't cut out)
const p4 = rect(startSketchOn('XY'), [60, 60], 50, 50)
  |> extrude(10, %)
const p4_1 = rect(startSketchOn(p3, "end"), [60, 45], 20, 20)
  |> extrude(-10, %)

// ❌ extrude with overhang (seems to work as expected [tested via export] but shows red lines)
const p6 = rect(startSketchOn('XY'), [0, -60], 50, 50)
  |> extrude(10, %)
const p6_1 = rect(startSketchOn(p6, "end"), [0, -40], 20, 20)
  |> extrude(10, %)

// ❌? extrude with no overlap (visually looks like two solids, exports as one)
const p7 = rect(startSketchOn('XY'), [60, -60], 50, 50)
  |> extrude(10, %)
const p7_1 = rect(startSketchOn(p7, "end"), [60, -100], 20, 20)
  |> extrude(10, %)

// ❌ negative extrude with overhang (shows red lines and does an xor-like thing, would expect either cut or join)
const p8 = rect(startSketchOn('XY'), [120, -60], 50, 50)
  |> extrude(10, %)
const p8_1 = rect(startSketchOn(p8, "end"), [120, -40], 20, 20)
  |> extrude(-10, %)

// ❌? negative adjacent extrude (shows red lines but seems to work)
const p9 = rect(startSketchOn('XY'), [180, -60], 50, 50)
  |> extrude(10, %)
const p9_1 = rect(startSketchOn(p9, "end"), [180, -95], 20, 20)
  |> extrude(-10, %)

// ❌ negative overlapping extrude (shows red lines and exports the edges)
const p10 = rect(startSketchOn('XY'), [240, -60], 50, 50)
  |> extrude(10, %)
const p10_1 = rect(startSketchOn(p10, "end"), [240, -60], 50, 50)
  |> extrude(-10, %)
```

**Comment by jessfraz** [comment]
assigning @jgomez720 to go thru and find out whether its an engine behavior, a real bug, or something where we are waiting on ssi

**Comment by jgomez720** [comment]
```js
// ❌ negative extrude through more than width of extrusion
const p5 = rect(startSketchOn('XY'), [120, 0], 50, 50)
  |> extrude(10, %)
const p5_1 = rect(startSketchOn(p5, "end"), [120, 0], 20, 20)
  |> extrude(-20, %)
```
Technically, working as intended. Engine does not supporting negative extrudes past the geometry. It should always assume cutting the existing geometry, and that's it. Subtract should solve this

```js
// ❌ extrude with edge coincident boundary this shows red lines
const p3 = rect(startSketchOn('XY'), [0, 60], 50, 50)
  |> extrude(10, %)
const p3_1 = rect(startSketchOn(p3, "end"), [0, 45], 20, 20)
  |> extrude(10, %)
```
I consider this a bug now. This works, so it shouldn't show red lines. Filed here:
https://github.com/KittyCAD/engine/issues/2130

```js
// ❌ negative extrude with edge coincident boundary (this doesn't cut out)
const p4 = rect(startSketchOn('XY'), [60, 60], 50, 50)
  |> extrude(10, %)
const p4_1 = rect(startSketchOn(p3, "end"), [60, 45], 20, 20)
  |> extrude(-10, %)
```
We will need subtract for this. If we implement a quick fix now, it's going to look similar to the first case above, where the extrusion passes the geometry, and still leaves behind artifacts.

```js
// ❌ extrude with overhang (seems to work as expected [tested via export] but shows red lines)
const p6 = rect(startSketchOn('XY'), [0, -60], 50, 50)
  |> extrude(10, %)
const p6_1 = rect(startSketchOn(p6, "end"), [0, -40], 20, 20)
  |> extrude(10, %)
```
Filed in the same issue above. This seems to work now, so the red lines are no longer needed. 

```js
// ❌? extrude with no overlap (visually looks like two solids, exports as one)
const p7 = rect(startSketchOn('XY'), [60, -60], 50, 50)
  |> extrude(10, %)
const p7_1 = rect(startSketchOn(p7, "end"), [60, -100], 20, 20)
  |> extrude(10, %)
```
Not entirely sure what to categorize this as, because I think I've seen multiple behaviors when you do something like this. The moment you create an extrude or revolve in OnShape that isn't connected to an existing body, it creates a new body entirely. I don't know how this is distinguished in a STEP file, but I have a feeling we aren't doing it the same way as OnShape is handling it.

```js
// ❌ negative extrude with overhang (shows red lines and does an xor-like thing, would expect either cut or join)
const p8 = rect(startSketchOn('XY'), [120, -60], 50, 50)
  |> extrude(10, %)
const p8_1 = rect(startSketchOn(p8, "end"), [120, -40], 20, 20)
  |> extrude(-10, %)
```
This is a bug. Should be fixed with SSI.

```js
// ❌? negative adjacent extrude (shows red lines but seems to work)
const p9 = rect(startSketchOn('XY'), [180, -60], 50, 50)
  |> extrude(10, %)
const p9_1 = rect(startSketchOn(p9, "end"), [180, -95], 20, 20)
  |> extrude(-10, %)
```
Filed in the same github issue above.

```js
// ❌ negative overlapping extrude (shows red lines and exports the edges)
const p10 = rect(startSketchOn('XY'), [240, -60], 50, 50)
  |> extrude(10, %)
const p10_1 = rect(startSketchOn(p10, "end"), [240, -60], 50, 50)
  |> extrude(-10, %)
```
Need SSI for this one.

**Comment by nadr0** [comment]
After discussing with Josh, he mentioned that this ticket is all engine issues.

**Comment by lf94** [comment]
Yes remove

**Comment by Irev-Dev** [comment]
if it's all engine issues should this be transferred to the engine?
